### PR TITLE
Adds gunmods that add slots for more gunmods

### PR DESCRIPTION
**Pull request by NotFuji**
#### Summary
```SUMMARY: Content "Adds gunmods that add slots for more gunmods"```  

#### Purpose of change
Some guns lack things like rails ( and some have more slots than they're supposed to ( mostly pistols ) ).  So this adds a system to add gunmods that add gunmod slots, enabling things such as scope mounts, railed handguards, and adding modernization kits like the M14-EBR without using an entirely new weapon.

#### Describe the solution
Gunmods are given a `add_mod` property which is added to its parent gun's mod slots when attached.

A `blacklist_mod` property is added to disallow a given mod on a gun that has that mod slot.  ( So a "rail" item, can be attachable to anything that doesn't already have a rail, for example. )

Removing the `add_mod` gunmod will also remove any mods that don't fit as a result.

An offset sight rail gunmod is added as an example.  It adds an additional sight slot.

### CODE REVIEW

**Comment by darktoes** [comment]
This seems really useful, and would let me have optional mod furniture for traditional bows. I approve.

File: src/item.cpp [hunk 1]
```diff
@@ -2431,14 +2468,32 @@ std::string item::info( std::vector<iteminfo> &info, const iteminfo_query *parts
     return format_item_info( info, {} );
 }
 
+std::map<gunmod_location, int> item::get_mod_locations() const {
```
Comment (AMurkin) [inline src/item.cpp]:
```suggestion
std::map<gunmod_location, int> item::get_mod_locations() const
{
```

File: src/item.cpp [hunk 2]
```diff
@@ -5276,6 +5331,12 @@ ret_val<bool> item::is_gunmod_compatible( const item &mod ) const
         return ret_val<bool>::make_failure( _( "must be unloaded before installing this mod" ) );
     }
 
+    for( auto slot : mod.type->gunmod->blacklist_mod ) {
+        if( get_mod_locations().count( slot ) ) {
+            return ret_val<bool>::make_failure( _( "cannot be installed on a weapon with \"%s\"" ), slot.name().c_str() );
```
Comment (AMurkin) [inline src/item.cpp]:
```suggestion
            return ret_val<bool>::make_failure( _( "cannot be installed on a weapon with \"%s\"" ),
                                                slot.name().c_str() );
```

**Comment by Night-Pryanik** [comment]
Please test if additional sight attached to the offset sight rail is properly removed from the weapon (and sometimes from the game) in case of weapon destruction (by fire, by explosion, by bashing, by unsuccessful attempt to attach a gunmod to a damaged weapon).

File: src/item.cpp [hunk 3]
```diff
@@ -2431,14 +2468,33 @@ std::string item::info( std::vector<iteminfo> &info, const iteminfo_query *parts
     return format_item_info( info, {} );
 }
 
+std::map<gunmod_location, int> item::get_mod_locations() const
```
Comment (AMurkin) [inline src/item.cpp]: Here is excessive space at the end.

**Pull request merged**